samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -59,27 +58,29 @@ class TaskInstance(
   streamMetadataCache: StreamMetadataCache = null,
   timerExecutor : ScheduledExecutorService = null,
   sideInputSSPs: Set[SystemStreamPartition] = Set(),
-  sideInputStorageManager: TaskSideInputStorageManager = null) extends Logging {
-
+  sideInputStorageManager: TaskSideInputStorageManager = null,
+  jobContext: JobContext,
+  containerContext: ContainerContext,
+  applicationContainerContext: Option[ApplicationContainerContext],
+  applicationTaskContextFactory: Option[ApplicationTaskContextFactory[ApplicationTaskContext]]
+) extends Logging {
+
+  val taskName: TaskName = taskModel.getTaskName
   val isInitableTask = task.isInstanceOf[InitableTask]
   val isWindowableTask = task.isInstanceOf[WindowableTask]
   val isEndOfStreamListenerTask = task.isInstanceOf[EndOfStreamListenerTask]
   val isClosableTask = task.isInstanceOf[ClosableTask]
   val isAsyncTask = task.isInstanceOf[AsyncStreamTask]
 
-  val kvStoreSupplier = ScalaJavaUtil.toJavaFunction(
-    (storeName: String) => {
-      if (storageManager != null && storageManager.getStore(storeName).isDefined) {
-        storageManager.getStore(storeName).get.asInstanceOf[KeyValueStore[_, _]]
-      } else if (sideInputStorageManager != null && sideInputStorageManager.getStore(storeName) != null) {
-        sideInputStorageManager.getStore(storeName).asInstanceOf[KeyValueStore[_, _]]
-      } else {
-        null
-      }
-    })
-
-  val context = new TaskContextImpl(taskName, metrics, containerContext, systemStreamPartitions.asJava, offsetManager,
-                                    kvStoreSupplier, tableManager, jobModel, streamMetadataCache, timerExecutor)
+  val epochTimeScheduler: EpochTimeScheduler = EpochTimeScheduler.create(timerExecutor)
+  val (taskContext: org.apache.samza.context.TaskContextImpl,
+    applicationTaskContextOption: Option[ApplicationTaskContext]) =
+    buildTaskContexts(epochTimeScheduler)
+  val context: Context = new ContextImpl(jobContext,
+    containerContext,
+    taskContext,
+    applicationContainerContext.orNull,
```

Comment:
Minor: Maybe rename to applicationContainerContextOption

Reply:
The type of the object already allows us to know it is an Option, and there isn't any variable in this class that uses the non-Option version, so I would prefer to be more concise here.

Reply:
I prefer suffixing Option/Future etc. because the type declaration can be far from the usage, so this makes it clearer when reviewing code or looking at it without an IDE.

Reply:
Ok, I'll update to Option/Optional.